File: asml/LC200NumberOfIslands.py
class Solution(object):
    def numIslands(self, grid):
        if not grid or not grid[0]:
            return 0
        answer = 0
        stack = []
        m = len(grid)
        n = len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    stack.append((i,j))
                    answer += 1
                    while stack:
                        row, col = stack.pop()
                        grid[row][col] = '2'
                        if row+1 <= m-1 and grid[row+1][col] == '1':
                            stack.append((row+1, col))
                        if row-1 >= 0 and grid[row-1][col] == '1':
                            stack.append((row-1, col))
                        if col+1 <= n-1 and grid[row][col+1] == '1':
                            stack.append((row, col+1))
                        if col-1 >= 0 and grid[row][col-1] == '1':
                            stack.append((row, col-1))
        return answer


        

File: asml/LC200NumberOfIslands.cpp
class Solution {
public:
    void BFS(vector<vector<char>> &grid, int i, int j){
        if(i>0&&grid[i-1][j]=='1'){
            grid[i-1][j]='0';
            BFS(grid, i-1, j);
        }
        if(j>0&&grid[i][j-1]=='1'){
            grid[i][j-1]='0';
            BFS(grid, i, j-1);
        }
        if(i<grid.size()-1&&grid[i+1][j]=='1'){
            grid[i+1][j]='0';
            BFS(grid, i+1, j);
        }
        if(j<grid[0].size()-1&&grid[i][j+1]=='1'){
            grid[i][j+1]='0';
            BFS(grid, i, j+1);
        }
    }
    int numIslands(vector<vector<char>> &grid) {
        int n=grid.size();
        if(n==0) return 0;
        int m=grid[0].size();

        int cnt=0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j]=='1'){
                    cnt++;
                    BFS(grid, i, j);
                }
            }
        }
        return cnt;
    }
};
